File: algebra_testbed/db/row.py
#
# Copyright (C) 2011 Akiban Technologies Inc.
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License, version 3,
# as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see http://www.gnu.org/licenses.
#

class Row(object):

    def __init__(self, rowtype, keyvals):
        self._rowtype = rowtype
        for k, v in keyvals.iteritems():
            self.__dict__[k] = v
        self._key = [keyvals[field] for field in rowtype.key]

    def __repr__(self):
        value = ['%s: %s' % (field, self.__dict__[field])
                 for field in self._rowtype.value]
        if len(self._key) > 0:
            return '%s%s: %s' % (self._rowtype, self._key, '{%s}' % ', '.join(value))
        else:
            return '%s: %s' % (self._rowtype, '{%s}' % ', '.join(value))

    def __cmp__(self, other):
        return cmp(self._key, other._key)

    def __getitem__(self, field):
        return self.__dict__[field]

    def __setitem__(self, field, value):
        self.__dict__[field] = value

    rowtype = property(lambda self: self._rowtype)

    key = property(lambda self: self._key)

    hkey = property(lambda self: self['hkey'])

    def ancestor_of(self, other):
        ancestor = False
        self_key = self.hkey
        other_key = other.hkey
        if self_key and other_key and len(self_key) <= len(other_key):
            ancestor = self_key == other_key[:len(self_key)]
        return ancestor

File: src/main/java/com/foundationdb/qp/row/Row.java
/**
 * Copyright (C) 2009-2013 FoundationDB, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.foundationdb.qp.row;

import com.foundationdb.ais.model.Table;
import com.foundationdb.server.types.value.ValueRecord;
import com.foundationdb.qp.rowtype.RowType;

public interface Row extends ValueRecord
{
    RowType rowType();
    HKey hKey();
    HKey ancestorHKey(Table table);
    boolean ancestorOf(Row that);
    boolean containsRealRowOf(Table table);
    Row subRow(RowType subRowType);
    boolean isBindingsSensitive();

    /**
     * Compares two rows and indicates if and where they differ.
     * @param row The row to be compared to this row.
     * @param leftStartIndex First field to compare in this row.
     * @param rightStartIndex First field to compare in the other row.
     * @param fieldCount Number of fields to compare.
     * @return 0 if all fields are equal. A negative value indicates that this row had the first field
     * that was not equal to the corresponding field in the other row. A positive value indicates that the
     * other row had the first field that was not equal to the corresponding field in this row. In both non-zero
     * cases, the absolute value of the return value is the position of the field that differed, starting the numbering
     * at 1. E.g. a return value of -2 means that the first fields of the rows match, and that in the second field,
     * this row had the smaller value.
     */
    int compareTo(Row row, int leftStartIndex, int rightStartIndex, int fieldCount);
}
